### ISSUE TITLE
problem of prediction

### ISSUE THREAD

**Issue by wwjwy** [qubvel/segmentation_models#456]
My dataset contains three categories, background, A and B. 
After I trained the model, I found that the number of pixel types in the prediction map exceeded three. Why?
Generally speaking, the number of classes is 3, shouldn’t the number of types of pixels in the prediction map also be three?
The code i use is as follows:

for i in ids:
    
    image, gt_mask = test_dataset[i]
    image = np.expand_dims(image, axis=0)
    pr_mask = model.predict(image)
    #cv2.imwrite("1.png",pr_mask*255)
    
    visualize(
        image=denormalize(image.squeeze()),
        gt_mask=gt_mask.squeeze(),
        pr_mask=pr_mask.squeeze(),
    )

**Comment by JordanMakesMaps** [comment]
Yes, the raw output should only have 3 classes (at most) if that's what it's trained on. You might want to check the raw output by doing something like:

```python
raw_prediction = model.predict(processed_image)
processed_prediction = np.argmax(raw_prediction.squeeze(), axis = 2)
print(np.unique(processed_prediction)) # 0, 1, and/or 2
```
I had a similar issue with the prediction/mask where I had values that weren't supposed to be there. I found out the reason was because when I resized the mask, I used `linear` or `cubic` interpolation, which can introduce new values. Now when I resize masks I always use `nearest_neighbor`.

**Comment by wwjwy** [comment]
> Yes, the raw output should only have 3 classes (at most) if that's what it's trained on. You might want to check the raw output by doing something like:
> 
> ```python
> raw_prediction = model.predict(processed_image)
> processed_prediction = np.argmax(raw_prediction.squeeze(), axis = 2)
> print(np.unique(processed_prediction)) # 0, 1, and/or 2
> ```
> 
> I had a similar issue with the prediction/mask where I had values that weren't supposed to be there. I found out the reason was because when I resized the mask, I used `linear` or `cubic` interpolation, which can introduce new values. Now when I resize masks I always use `nearest_neighbor`.

Thanks you very much!
I have another question, which wants to get your help.
In my three categories (background, A, B), the background pixel ratio is very high. When I train, it is necessary to set the loss function class_indexes=[0,1] (ignore the effect of the background on the loss function )?

**Comment by JordanMakesMaps** [comment]
For classes ['Background', 'A', B'] you can train on all of them, and then when testing occurs, ignore pixels that correspond to to the 'Background' class. Or, you can ignore 'Background' by zeroing the index in the loss function while keep the indices for 'A' and 'B' non-zero. Doing this means that the model will not learn the 'Background' class, which may be ideal, but it also means that when you provide images during the testing time, the entire image will be predicted either 'A' or 'B', even if 'Background' is present (it won't ignore it, though you might be able to threshold it out).

```python
loss = sm.losses.some_loss_function(class_indexes = [0, 1, 2], class_weights ] = [0.0, 1.0, 1.0]) # here we're ignoring 'Background'
```

**Comment by wwjwy** [comment]
> For classes ['Background', 'A', B'] you can train on all of them, and then when testing occurs, ignore pixels that correspond to to the 'Background' class. Or, you can ignore 'Background' by zeroing the index in the loss function while keep the indices for 'A' and 'B' non-zero. Doing this means that the model will not learn the 'Background' class, which may be ideal, but it also means that when you provide images during the testing time, the entire image will be predicted either 'A' or 'B', even if 'Background' is present (it won't ignore it, though you might be able to threshold it out).
> 
> ```python
> loss = sm.losses.some_loss_function(class_indexes = [0, 1, 2], class_weights ] = [0.0, 1.0, 1.0]) # here we're ignoring 'Background'
> ```

Thank you very much!
